src/providers/ibmcloud/gen2.rs

```diff
@@ -0,0 +1,196 @@
+//! Metadata fetcher for IBMCloud (VPC Gen2) instances.
+//!
+//! This provider supports the "VPC Generation 2" infrastructure type
+//! on IBMCloud.
+//! It provides a config-drive as the only metadata source, whose layout
+//! is very similar to `cloud-init NoCloud` [datasource][nocloud], with
+//! a few variations:
+//!  - disk label is `cidata` (lowercase)
+//!  - filesystem is `iso9660`
+//!
+//! nocloud: https://cloudinit.readthedocs.io/en/latest/topics/datasources/nocloud.html
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{BufRead, BufReader, Read};
+use std::path::{Path, PathBuf};
+
+use openssh_keys::PublicKey;
+use slog_scope::{error, warn};
+use tempdir::TempDir;
+
+use crate::errors::*;
+use crate::network;
+use crate::providers::MetadataProvider;
+
+const CONFIG_DRIVE_LABEL: &str = "cidata";
+
+/// IBMCloud provider (VPC Gen2).
+#[derive(Debug)]
+pub struct G2Provider {
+    /// Path to the top directory of the mounted config-drive.
+    drive_path: PathBuf,
+    /// Temporary directory for own mountpoint (if any).
+    temp_dir: Option<TempDir>,
+}
+
+impl G2Provider {
+    /// Try to build a new provider client.
+    ///
+    /// This internally tries to mount (and own) the config-drive.
+    pub fn try_new() -> Result<Self> {
+        let target =
+            TempDir::new("afterburn").chain_err(|| "failed to create temporary directory")?;
+        crate::util::mount_ro(
+            &Path::new("/dev/disk/by-label/").join(CONFIG_DRIVE_LABEL),
+            target.path(),
+            "iso9660",
+            3,
+        )?;
+
+        let provider = Self {
+            drive_path: target.path().to_owned(),
+            temp_dir: Some(target),
+        };
+        Ok(provider)
+    }
+
+    /// Return the path to the metadata directory.
+    fn metadata_dir(&self) -> PathBuf {
+        self.drive_path.clone()
+    }
+
+    /// Read metadata file and parse attributes.
+    fn read_metadata(&self) -> Result<HashMap<String, String>> {
+        let filename = self.metadata_dir().join("meta-data");
+        let file =
+            File::open(&filename).chain_err(|| format!("failed to open file '{:?}'", filename))?;
+        let bufrd = BufReader::new(file);
+        Self::parse_metadata(bufrd)
+    }
+
+    /// Parse metadata attributes.
+    ///
+    /// Metadata file contains one attribute per line, in the form of
+    /// `key: value\n`.
+    fn parse_metadata<T: Read>(input: BufReader<T>) -> Result<HashMap<String, String>> {
+        let mut output = HashMap::new();
+
+        for line in input.lines().filter_map(|l| l.ok()) {
+            let parts: Vec<&str> = line.splitn(2, ':').collect();
```

Comment:
`splitn` would always return 2 parts, wouldn't it? In this case the following `parts.len()` is useless.

Perhaps just `split` should be used instead - do we how IBMCloud treats lines like `foo:bar:baz`?

Reply:
Not always; it returns at most 2 parts, see https://doc.rust-lang.org/std/primitive.str.html#method.splitn.

I don't have any way to test the "embedded :" case, as I can only directly test/control the hostname case and their API reject any name with a `:` inside (with `Use lowercase alphanumeric characters and hyphens only (without spaces)`).

Reply:
Not always; it returns at most 2 parts, see https://doc.rust-lang.org/std/primitive.str.html#method.splitn.

I don't have any way to test the "embedded :" case, as I can only directly test/control the hostname case and their API reject any name with a `:` inside (with `Use lowercase alphanumeric characters and hyphens only (without spaces)`).

Reply:
Ah, that's protecting from `foo:` case, got it.

Lets leave it like it is then